packages/web3-utils/src/validation.ts

```diff
@@ -162,8 +153,22 @@ export const compareBlockNumbers = (blockA: Numbers, blockB: Numbers) => {
 	if (blockB === 'pending') {
 		return -1;
 	}
+	if (blockA === 'safe' && blockB === 0) {
+		return 1;
+	}
+	if (blockA === 0 && blockB === 'safe') {
+		return -1;
+	}
+	if (blockA === 'safe' || blockB === 'safe') {
+		// either a or b is "safe" and the other one did not fall into any of the conditions above, so the other one is a number
+		throw new InvalidBlockError(
+			`Cannot compare safe tag with ${blockA === 'safe' ? blockB : blockA}`,
+		);
+	}
```

Comment:
> I think, any block-tag vs block-num cant be compared as the block number provided can be on the left or the right side of that block tag, until that tag's block number is extracted and compared.

in some cases we can compare tag with number as:
earliest vs num
pending vs num, ( as pending means it will always come after mined block )

Reply:
https://github.com/web3/web3.js/issues/5836